text/0193-fixing-type-unions.md

```diff
@@ -0,0 +1,558 @@
+  }
+  ```
+
+- In **Go**:
+
+  > :construction: TBD
+
+- In **Java**:
+
+  ```java
+  // UnionType.java
+  public final class UnionType {
+    public static UnionType fromFoo(Foo value) { /* ... */ }
+    public static UnionType fromMapOfBar(Map<String, Bar> value) { /* ... */ }
+    public static UnionType fromListOfBaz(Baz... value) { /* ... */ }
+    public static UnionType fromListOfBaz(List<Baz> value) { /* ... */ }
+
+    public @Nullable Foo asFoo() { /* ... */ }
+    public @Nullable Map<String, Bar> asMapOfBar() { /* ... */ }
+    public @Nullable List<Baz> asListOfBaz() { /* ... */ }
+  }
+  ```
+
+The _jsii Kernel_ will need adjustments in order to better support those types. The generated libraries will provide a
+new argument to the _jsii kernel_'s `load` API if it's code was generated with `--union-like-types` enabled. This way,
+the kernel will be bale to reference the declared _type union_'s information when annotating instances that are passed
+from **Javascript** to the host application, removing all ambiguity regarding which type is serialized.
+
+# Drawbacks
+
+This approach might be expensive to implement as it might require fundamentally changing how the `jsii` compiler works.
+It currently operates on the `tsc` _Type Checker_ API, which preprocesses type aliases away in order to have a more
+streamlined type system to reason over. In order to property process aliased _type unions_, it may be necessary to fall
+back on the `Node` API, which allows more direct inspection of the original source code, however provides fewer
+inspection capabilities (since the type system has not been fully resolved at the time a `Node` is processed).
+
+# Rationale and Alternatives
+
+Improving support for _type unions_ appears to be the best way going forward, as complete removal of type unions has
+proven to be challenging from an API design pespective: it tends to make the experience significantly worse for dynamic
+languages (as boilerplate-heavy patterns must be leveraged as a replacement for _type unions_), and would reduce the
+flexibility in language-specific code generation for _union-like_ types (as those would be bound to the **TypeScript**
+API). Full support of _type unions_ as a type kind in the _jsii type model_ achieves better outcomes:
+
+- it ensures consistency around _union-like_ types
+- it does not degrade the developer experience in languages with _native type_ union support
+- it requires significantly less effort from library authors
+
+This feature could however not be introduced as part of CDK v1, as the change of static (and dynamic) type it requires
+in statically typed languages would be breaking to existing customers. The release of [AWS CDK v2] is a prime
+opportunity to introduce this change and to address a well-known customer pain point.
+
+# Adoption Strategy
+
+This feature is a candidate for rollout as part of [AWS CDK v2]. Since the removal of type unions will incur significant
```

Comment:
I am wondering if we should wait for 3.x to "flip the switch" because there is quite a lot of work over the 1.x line (both CDK and jsii) that can be done before we are ready to introduce the breaking change to the CDK.

I would recommend to kick-off the non-breaking work on jsii and CDK and then flip the switch in the next CDK major release after this code has been baked and used on the 1.x lines (of both projects) for a few months.

Reply:
Oh looks like this is a leftover from the previous iteration. The intention is quite a bit different now :)